Frontend reports

WesleySkillern opened the pull request:
### Please give a short summary of what's included in this submission
This rounds out the rest of the Reports for MVP. It also adds a title and default chart to Reports as requested by Nico. 

It also changes the H2MS banner link to route to the Results page. A background image is added as per Michael's request (and credited in About). 

### Checklist
- [x] The Pull Request title describes the changes I've made
- [x] I've reviewed the **Files changed** tab
- [x] The code in the **Files changed** tab meets our style guides
- [x] New classes and public methods in the **Files changed** tab are documented
- [x] I've added two reviewers for this pull request


### Frontend Reviewer Checklist
##### I've reviewed the observe page, reports page, and any page directly affected by these changes
- [x] For Mobile (width: ~320px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=320&h=480)
- [x] For Tablet landscape (width: ~1024px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=1024&h=768)
- [x] For Desktop (width: 1920px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=1920&h=1080)
- [x] I've reviewed the **Files Changed** Tab and left constructive criticism (if applicable)

###### Optional
- [ ] For Tablet portrait (width: ~768px) [View](http://whatismyscreenresolution.net/multi-screen-test?site-url=http://localhost:4200/login&w=768&h=1024)

benjenkinsv95 commented on frontend/src/app/app.component.css:
```diff
@@ -22,6 +22,8 @@ h1.app-name {
     /* When the sidenav is not fixed, stretch the sidenav container to fill the available space. This
        causes `<mat-sidenav-content>` to act as our scrolling element for desktop layouts. */
     flex: 1;
+    background: url('../assets/images/backgrounds/h2ms-background.jpg');
```
`..` isn't needed. You can start at `/assets`

benjenkinsv95 commented on frontend/src/app/config/local-h2ms-config.ts:
```diff
@@ -9,4 +9,5 @@ export const CONFIG: Config = new Config('H2MS',
     'http://localhost',
-    8080);
+    8080,
+    [{id: null, displayName: null}]);
```
Should be the same as the live `h2ms-config.ts` file.

Ex.
`[{id: '3', displayName: 'Washed?'}]);`

benjenkinsv95 commented on frontend/src/app/reports/reports-chart.service.ts:
```diff
@@ -0,0 +1,154 @@
+    chart: ChartAPI;
+
+    constructor() {
+    }
```
Unneeded.

benjenkinsv95 commented on frontend/src/app/reports/reports-chart.service.ts:
```diff
@@ -0,0 +1,154 @@
+import {Injectable} from '@angular/core';
+import * as c3 from 'c3';
+import {ChartAPI} from 'c3';
```
This kind of looks off to me. You've already imported all of c3, but then you re-import the ChartAPI, when it should already be available.

WesleySkillern replied:
ya, but necessary...

benjenkinsv95 commented on frontend/src/app/reports/reports-chart.service.ts:
```diff
@@ -0,0 +1,154 @@
+        }
+        let number = 0;
+        if (names.length > 5) {
+            number = -70;
```
`number` should be renamed to show intent. It looks like it has something to do with rotation. So maybe `degrees`?

benjenkinsv95 commented on frontend/src/app/reports/reports-chart.service.ts:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    /**
+     * a helder function to convert a number to its ordinal (1st, 2nd, etc.) string value
```
"helder function"

benjenkinsv95 commented on frontend/src/app/reports/reports-chart.service.ts:
```diff
@@ -0,0 +1,154 @@
+    /**
+     * a helder function to convert a number to its ordinal (1st, 2nd, etc.) string value
+     */
+    getGetOrdinal(n: number): string {
```
Not enough "get" I recommend renaming to `getGetGetGetGetOrdinal`

benjenkinsv95 commented on frontend/src/app/reports/reports.component.ts:
```diff
@@ -46,31 +34,41 @@ export class ReportsComponent implements OnInit {
+        this.progressBarIsHidden = false;
         this.emptyJSONReturned = false;
+        this.charts = this.reportsService.getCharts();
+        // defaults
```
Not sure how much value the comment provides. Constructors and "onInit" functions commonly set default values.

benjenkinsv95 commented on frontend/src/app/reports/reports.service.ts:
```diff
@@ -40,27 +69,98 @@ export class ReportsService {
+                /events/count/observer
+     */
+    private buildURL(chart, grouping) {
+        if (chart.value.match('Number of observations')) {
```
This function is hard to follow. Might try reducing the if/else branches, check this out on how: https://youtu.be/ldqDpmMkXgw or break chunks out into functions that define intent.

benjenkinsv95 commented on frontend/src/index.html:
```diff
@@ -4,7 +4,7 @@
     <base href="/">
     <!-- todo: move favicon.ico to config -->
-    <link rel="icon" type="image/x-icon" href="favicon.ico">
+    <link rel="icon" type="image/x-icon" href="./assets/images/favicons/h2ms-favicon.ico">
```
Don't think `.` is needed as a prefix

benjenkinsv95 commented:
Reports page awkwardly has horizontal bar on tablet views: ![](https://i.imgur.com/rzDBD9g.png)

I don't know when this was introduced though.

benjenkinsv95 commented:
Horizontal bar also occurs on login page on landscape tablet ![](https://i.imgur.com/TW3fBoo.png)

WesleySkillern commented:
@benjenkinsv95 Thanks for all your comments. Commits should've addressed most of your concerns (e.g. consistent capitalization, trailing decimals, etc.). I'm not sure if simple if-else inversion will help all the cases for plots. 

Also, not sure about the horizontal scroll bar. I don't have it for Chrome or Safari.

Other than that, it should be good. Please let me know if there's anything else, or anything I missed.
